src/mito2/src/manifest/action.rs

```diff
@@ -196,8 +260,72 @@ impl RegionManifestBuilder {
     }
 }
 
+/// A record of removed files in the region manifest.
+/// This is used to keep track of files that have been removed from the manifest but may still
+/// be kept for a while
+#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq)]
+pub struct RemovedFilesRecord {
+    /// a list of `(FileIds, timestamp)` pairs, where the timestamp is the time when
+    /// the files are removed from manifest. The timestamp is in milliseconds since unix epoch.
+    pub removed_files: Vec<RemovedFiles>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq)]
+pub struct RemovedFiles {
+    /// The timestamp is the time when
+    /// the files are removed from manifest. The timestamp is in milliseconds since unix epoch.
+    pub removed_at: i64,
+    /// The set of file ids that are removed.
+    pub file_ids: HashSet<FileId>,
+}
+
+impl RemovedFilesRecord {
+    /// Add a record of removed files with the current timestamp.
+    pub fn add_removed_files(&mut self, file_ids: HashSet<FileId>, at: i64) {
+        self.removed_files.push(RemovedFiles {
+            removed_at: at,
+            file_ids,
+        });
+    }
+
+    pub fn evict_old_removed_files(&mut self, opt: &RemoveFileOptions) -> Result<()> {
+        let total_removed_files: usize = self.removed_files.iter().map(|s| s.file_ids.len()).sum();
+        if total_removed_files <= opt.keep_count {
+            return Ok(());
+        }
+
+        let mut cur_file_cnt = total_removed_files;
+
+        let can_evict_until = chrono::Utc::now()
+            - chrono::Duration::from_std(opt.keep_ttl).context(DurationOutOfRangeSnafu {
+                input: opt.keep_ttl,
+            })?;
+
+        self.removed_files.sort_unstable_by_key(|f| f.removed_at);
+        let updated = std::mem::take(&mut self.removed_files)
+            .into_iter()
+            .filter_map(|f| {
+                if f.removed_at < can_evict_until.timestamp_millis()
+                    && cur_file_cnt >= opt.keep_count
+                {
+                    // can evict all files
+                    // TODO(discord9): maybe only evict to below keep_count? Maybe not, or the update might be too frequent.
+                    cur_file_cnt -= f.file_ids.len();
+                    None
+                } else {
+                    Some(f)
+                }
+            })
+            .collect();
+        self.removed_files = updated;
```

Comment:
I think it would be better to hard limit the file num under keep_count otherwise this option is meaningless. We can take the last keep_count elements in `updated`.